core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java

```diff
@@ -191,7 +221,78 @@ public AtomicReference<Long> getWorkerId() {
 
   @Override
   public FileInfo getFileInfo(String fileId) throws IOException {
-    return new FileInfo();
+    alluxio.grpc.FileInfo fi;
+    String ufsFullPath = PathUtils.concatPath(mRootUFS, fileId);
+    String fn = new AlluxioURI(fileId).getName();
+
+    UfsFileStatus status = mUfsFileStatusCache.getIfPresent(ufsFullPath);
+    if (status == null) {
+      // The requested FileStatus is not present in memory cache.
+      // Let's try to query local persistent DoraMetaStore.
+      Optional<DoraMeta.FileStatus> fs = mMetaStore.getDoraMeta(ufsFullPath);
+      if (fs.isPresent()) {
+        // Found in persistent DoraMetaStore
+        fi = fs.get().getFileInfo();
+        String contentHash = UnderFileSystemUtils.approximateContentHash(fi.getLength(),
+            fi.getLastModificationTimeMs());
+        UfsFileStatus ufs = new UfsFileStatus(fi.getPath(), contentHash, fi.getLength(),
+            fi.getLastModificationTimeMs(),
+            fi.getOwner(), fi.getGroup(), (short) fi.getMode(), fi.getBlockSizeBytes());
+        mUfsFileStatusCache.put(ufsFullPath, ufs);
+      } else {
+        // This will load UfsFileStatus from UFS and put it in memory cache
+        try {
+          status = mUfsFileStatusCache.get(ufsFullPath);
+        } catch (ExecutionException e) {
+          Throwable throwable = e.getCause();
+          // this should be the exception thrown by ufs.getFileStatus which is IOException
+          if (throwable instanceof IOException) {
+            throw (IOException) throwable;
+          } else {
+            throw new InternalException("Unexpected exception when retrieving UFS file status",
+                throwable);
+          }
+        }
+        fi = buildFileInfoFromUfsFileStatus(status, fn, fileId, ufsFullPath);
+
+        // Add this to persistent DoraMetaStore.
+        long currentTimeMillis = System.currentTimeMillis();
+        mMetaStore.putDoraMeta(ufsFullPath,
+            DoraMeta.FileStatus.newBuilder().setFileInfo(fi).setTs(currentTimeMillis).build());
+      }
+    } else {
+      // Found in memory cache
+      fi = buildFileInfoFromUfsFileStatus(status, fn, fileId, ufsFullPath);
+    }
+    // because cache manager uses hased ufs path as file ID
+    // TODO(bowen): we need a dedicated type for file IDs!
+    String cacheManagerFileId = new AlluxioURI(ufsFullPath).hash();
+
+    long bytesInCache = mCacheManager.getUsage()
+        .flatMap(usage -> usage.partitionedBy(file(cacheManagerFileId)))
+        .map(CacheUsage::used).orElse(0L);
+    int cachedPercentage = (int) (bytesInCache / fi.getLength()) * 100;
```

Comment:
fi.getLength() may be 0

Reply:
Fixed in [61e9a78](https://github.com/Alluxio/alluxio/pull/16879/commits/61e9a78772d16f8378cf41d31c42fa1a1fdb50ba)